### PR TITLE
fix(VTreeview): disabled items should be ignored when selecting parent

### LINKED ISSUE #8244: [Bug Report] Treeview: It's possible to select a disabled item by selecting its parent
### Environment
**Vuetify Version:** 2.0.3
**Vue Version:** 2.6.10
**Browsers:** Chrome 77.0.3853.0, Google Chrome
**OS:** Windows 10

### Steps to reproduce
Set a child item as disabled in a selectable treeview
Click on its parent node

### Expected Behavior
The disabled item state should not change

### Actual Behavior
The disabled item state is changing

### Reproduction Link
<a href="https://codepen.io/anon/pen/NQgVwR" target="_blank">https://codepen.io/anon/pen/NQgVwR</a>

### PR DESCRIPTION
**Pull request by nekosaur**
## Description
selecting parent item did not respect disabled children

## Motivation and Context
closes #8244

## How Has This Been Tested?

playground, unit test

## Markup:

<details>

```vue
<template>
  <v-content>
    <v-container grid-list-xl>
      <v-treeview
        v-model="selected"
        :items="items"
        selectable
      />
      {{ selected }}
    </v-container>
  </v-content>
</template>

<script>
  export default {
    data: () => ({
      selected: [4],
      items: [
        {
          id: 1,
          name: 'Parent',
          children: [
            {id: 2, name: 'Child 1'},
            {id: 3, name: 'Child 2'},
            {id: 4, name: 'Child 3', disabled: true },
            {id: 5, name: 'Child 4', disabled: true},
          ]
        }
      ],
    }),
  }
</script>

```
</details>

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 1]
```diff
@@ -157,7 +157,7 @@ export default mixins(
 
   created () {
     this.buildTree(this.items)
-    this.value.forEach(key => this.updateSelected(this.returnObject ? getObjectValueByPath(key, this.itemKey) : key, true))
+    this.value.forEach(key => this.updateSelected(this.returnObject ? getObjectValueByPath(key, this.itemKey) : key, true, true))
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:160]: break into new lines please

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 2]
```diff
@@ -322,19 +322,24 @@ export default mixins(
...
+        this.getDescendants(key).forEach(descendant => {
+          if (!getObjectValueByPath(this.nodes[descendant].item, this.itemDisabled) || isForced) {
+            this.nodes[descendant].isSelected = isSelected
+            this.nodes[descendant].isIndeterminate = false
+            changed.set(descendant, isSelected)
+          }
         })
 
+        this.nodes[key].isSelected = isSelected
+        this.nodes[key].isIndeterminate = this.getDescendants(key).some(descendant => !this.nodes[descendant].isSelected)
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:340]: you and your long lines

**Comment by dinhtq** [comment]
I think this fix mostly works, although I think there is one edge case that it fails. 
The edge case is when there is only **unselected disabled child node** along with enabled nodes, the parent node should be either indeterminate or fully unselected.
Screenshot 2020-04-28 at 2 15 12 PM

**Comment by dinhtq** [comment]
I modified your fixes into another PR with an additional expression 
https://github.com/vuetifyjs/vuetify/pull/11260

**Comment by nekosaur** [comment]
@dinhtq What steps did you take to get the treeview in that state?

**Comment by dinhtq** [comment]
anyway we can get this fix expedited?

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 3]
```diff
@@ -157,8 +157,9 @@ export default mixins(
 
   created () {
     this.buildTree(this.items)
-    this.value.forEach(key => this.updateSelected(this.returnObject ? getObjectValueByPath(key, this.itemKey) : key, true))
-    this.active.forEach(key => this.updateActive(this.returnObject ? getObjectValueByPath(key, this.itemKey) : key, true))
+    const getValue = (key: string | number) => this.returnObject ? getObjectValueByPath(key, this.itemKey) : key
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:160]: change ternary to 3 lines

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 4]
```diff
@@ -157,8 +157,9 @@ export default mixins(
...
+    this.value.map(getValue).forEach(value => this.updateSelected(value, true, true))
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:161]: use for..of instead

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 5]
```diff
@@ -157,8 +157,9 @@ export default mixins(
...
+    this.active.map(getValue).forEach(value => this.updateActive(value, true))
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:162]: use for..of instead

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 6]
```diff
@@ -222,7 +223,13 @@ export default mixins(
         node.isActive = oldNode.isActive
         node.isOpen = oldNode.isOpen
 
-        this.nodes[key] = !children.length ? node : this.calculateState(node, this.nodes)
+        this.nodes[key] = node
+
+        if (children.length) {
+          const { isSelected, isIndeterminate } = this.calculateState(key, this.nodes)
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:229]: blank line after variable declarations

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 7]
```diff
@@ -232,17 +239,20 @@ export default mixins(
         this.updateVnodeState(key)
       }
     },
-    calculateState (node: NodeState, state: Record<string | number, NodeState>) {
-      const counts = node.children.reduce((counts: number[], child: string | number) => {
+    calculateState (node: string | number, state: Record<string | number, NodeState>) {
+      const counts = state[node].children.reduce((counts: number[], child: string | number) => {
         counts[0] += +Boolean(state[child].isSelected)
         counts[1] += +Boolean(state[child].isIndeterminate)
         return counts
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:246]: blank line before return

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 8]
```diff
@@ -232,17 +239,20 @@ export default mixins(
...
       }, [0, 0])
 
-      node.isSelected = !!node.children.length && counts[0] === node.children.length
-      node.isIndeterminate = !node.isSelected && (counts[0] > 0 || counts[1] > 0)
+      const isSelected = !!state[node].children.length && counts[0] === state[node].children.length
+      const isIndeterminate = !isSelected && (counts[0] > 0 || counts[1] > 0)
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:250]: group this with parenthesis

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 9]
```diff
@@ -322,23 +332,31 @@ export default mixins(
...
 
       const changed = new Map()
 
       if (this.selectionType !== 'independent') {
-        const descendants = [key, ...this.getDescendants(key)]
-        descendants.forEach(descendant => {
-          this.nodes[descendant].isSelected = isSelected
-          this.nodes[descendant].isIndeterminate = false
-          changed.set(descendant, isSelected)
+        this.getDescendants(key).forEach(descendant => {
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:341]: use for..of instead

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 10]
```diff
@@ -322,23 +332,31 @@ export default mixins(
...
+          if (!getObjectValueByPath(this.nodes[descendant].item, this.itemDisabled) || isForced) {
+            this.nodes[descendant].isSelected = isSelected
+            this.nodes[descendant].isIndeterminate = false
+            changed.set(descendant, isSelected)
+          }
         })
 
+        this.nodes[key].isSelected = isSelected
+        const calculated = this.calculateState(key, this.nodes)
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:350]: don't smush variables between logic. define as high as possible. same guidance as above.

File: packages/vuetify/src/components/VTreeview/VTreeview.ts [hunk 11]
```diff
@@ -322,23 +332,31 @@ export default mixins(
...
+        this.nodes[key].isIndeterminate = calculated.isIndeterminate
+        changed.set(key, isSelected)
+
         const parents = this.getParents(key)
         parents.forEach(parent => {
-          this.nodes[parent] = this.calculateState(this.nodes[parent], this.nodes)
-          changed.set(parent, this.nodes[parent].isSelected)
+          const calculated = this.calculateState(parent, this.nodes)
```
Comment (johnleider) [inline packages/vuetify/src/components/VTreeview/VTreeview.ts:356]: don't smush variables between logic. define as high as possible. same guidance as above.

File: packages/vuetify/src/components/VTreeview/VTreeview.ts (hunk 3 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:160]: not sure what you mean here

File: packages/vuetify/src/components/VTreeview/VTreeview.ts (hunk 8 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:250]: not sure what you mean here

File: packages/vuetify/src/components/VTreeview/VTreeview.ts (hunk 3 above)
Comment (johnleider) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:160]: the single line is long (subjective). disregard if you disagree

File: packages/vuetify/src/components/VTreeview/VTreeview.ts (hunk 8 above)
Comment (johnleider) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:250]:
```js
!!state[node].children.length && (counts[0] === state[node].children.length)
```

https://www.notion.so/vuetify/General-8ef7db0ab62a49f3a2f7cec7b67f3a44#d10855e35ea742158c874e5f08cf02ba

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:250]: how does it apply here? it's neither case 1 (`cb && cb()`) nor case 2 (`this.showWeek && day < daysInMonth`)

Comment (johnleider) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:250]: Do we need to have an example for every possible scenario?

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:250]: oh, you tagged the wrong line

File: packages/vuetify/src/components/VTreeview/VTreeview.ts (hunk 10 above)
Comment (nekosaur) [reply packages/vuetify/src/components/VTreeview/VTreeview.ts:350]: it has to be smushed because of logic

**Comment by SirPoot-** [comment]
Thank you for this fix! Extremely helpful and can not wait to implement it.

**Pull request merged**